File: c/snake.c
/*
 * Snake Game — C (C99)
 * Terminal-based using ncurses.
 * Compile: gcc -o snake snake.c -lncurses
 * Run:     ./snake
 */

#include <ncurses.h>
#include <stdlib.h>
#include <time.h>

#define WIDTH  30
#define HEIGHT 20
#define MAX_LEN (WIDTH * HEIGHT)

typedef struct { int x, y; } Vec2;

typedef struct {
    Vec2 body[MAX_LEN];
    int  length;
    Vec2 dir;
    Vec2 food;
    int  score;
    int  alive;
} Game;

static void place_food(Game *g) {
    int occupied;
    do {
        g->food.x = rand() % WIDTH;
        g->food.y = rand() % HEIGHT;
        occupied = 0;
        for (int i = 0; i < g->length; i++) {
            if (g->body[i].x == g->food.x && g->body[i].y == g->food.y) {
                occupied = 1;
                break;
            }
        }
    } while (occupied);
}

static void init_game(Game *g) {
    g->length = 3;
    g->score  = 0;
    g->alive  = 1;
    g->dir    = (Vec2){1, 0};

    int mx = WIDTH / 2, my = HEIGHT / 2;
    g->body[0] = (Vec2){mx, my};
    g->body[1] = (Vec2){mx - 1, my};
    g->body[2] = (Vec2){mx - 2, my};

    place_food(g);
}

static void tick(Game *g) {
    if (!g->alive) return;

    Vec2 head = {
        g->body[0].x + g->dir.x,
        g->body[0].y + g->dir.y
    };

    /* Wall check */
    if (head.x < 0 || head.x >= WIDTH || head.y < 0 || head.y >= HEIGHT) {
        g->alive = 0;
        return;
    }

    /* Self check */
    for (int i = 0; i < g->length; i++) {
        if (g->body[i].x == head.x && g->body[i].y == head.y) {
            g->alive = 0;
            return;
        }
    }

    int ate = (head.x == g->food.x && head.y == g->food.y);

    /* Shift body */
    if (!ate) {
        for (int i = g->length - 1; i > 0; i--)
            g->body[i] = g->body[i - 1];
    } else {
        g->score += 10;
        for (int i = g->length; i > 0; i--)
            g->body[i] = g->body[i - 1];
        g->length++;
        place_food(g);
    }

    g->body[0] = head;
}

static void draw(const Game *g) {
    erase();
    int h, w;
    getmaxyx(stdscr, h, w);
    int ox = (w - WIDTH * 2 - 2) / 2;
    int oy = (h - HEIGHT - 4) / 2;
    if (ox < 0) ox = 0;
    if (oy < 0) oy = 0;

    /* Score */
    attron(COLOR_PAIR(4) | A_BOLD);
    mvprintw(oy, ox, " SCORE: %d ", g->score);
    attroff(COLOR_PAIR(4) | A_BOLD);

    /* Border */
    attron(COLOR_PAIR(3));
    mvaddch(oy + 1, ox, '+');
    mvaddch(oy + 1, ox + WIDTH * 2 + 1, '+');
    mvaddch(oy + HEIGHT + 2, ox, '+');
    mvaddch(oy + HEIGHT + 2, ox + WIDTH * 2 + 1, '+');
    for (int x = 1; x <= WIDTH * 2; x++) {
        mvaddch(oy + 1, ox + x, '-');
        mvaddch(oy + HEIGHT + 2, ox + x, '-');
    }
    for (int y = 0; y < HEIGHT; y++) {
        mvaddch(oy + 2 + y, ox, '|');
        mvaddch(oy + 2 + y, ox + WIDTH * 2 + 1, '|');
    }
    attroff(COLOR_PAIR(3));

    /* Food */
    attron(COLOR_PAIR(2) | A_BOLD);
    mvprintw(oy + 2 + g->food.y, ox + 1 + g->food.x * 2, "@@");
    attroff(COLOR_PAIR(2) | A_BOLD);

    /* Snake */
    for (int i = 0; i < g->length; i++) {
        attron(COLOR_PAIR(1) | (i == 0 ? A_BOLD : 0));
        mvprintw(oy + 2 + g->body[i].y, ox + 1 + g->body[i].x * 2,
                 i == 0 ? "##" : "oo");
        attroff(COLOR_PAIR(1) | (i == 0 ? A_BOLD : 0));
    }

    /* Game over */
    if (!g->alive) {
        attron(A_REVERSE | A_BOLD);
        mvprintw(oy + 2 + HEIGHT / 2, ox + 2,
                 "  GAME OVER  Score: %d  'r'=restart 'q'=quit  ", g->score);
        attroff(A_REVERSE | A_BOLD);
    }

    refresh();
}

int main(void) {
    srand((unsigned)time(NULL));

    initscr();
    cbreak();
    noecho();
    keypad(stdscr, TRUE);
    curs_set(0);
    timeout(110);
    start_color();
    use_default_colors();

    init_pair(1, COLOR_GREEN,  -1);
    init_pair(2, COLOR_RED,    -1);
    init_pair(3, COLOR_CYAN,   -1);
    init_pair(4, COLOR_YELLOW, -1);

    Game game;
    init_game(&game);

    while (1) {
        draw(&game);
        int ch = getch();

        if (ch == 'q') break;

        if (!game.alive) {
            if (ch == 'r') init_game(&game);
            continue;
        }

        switch (ch) {
            case KEY_UP:    case 'w': if (game.dir.y != 1)  game.dir = (Vec2){ 0, -1}; break;
            case KEY_DOWN:  case 's': if (game.dir.y != -1) game.dir = (Vec2){ 0,  1}; break;
            case KEY_LEFT:  case 'a': if (game.dir.x != 1)  game.dir = (Vec2){-1,  0}; break;
            case KEY_RIGHT: case 'd': if (game.dir.x != -1) game.dir = (Vec2){ 1,  0}; break;
        }

        tick(&game);
    }

    endwin();
    return 0;
}

File: python/snake.py
"""
Snake Game — Python 3
Terminal-based using curses library.
Run: python snake.py
"""

import curses
import random
from collections import deque
from enum import Enum


class Direction(Enum):
    UP = (0, -1)
    DOWN = (0, 1)
    LEFT = (-1, 0)
    RIGHT = (1, 0)


OPPOSITE = {
    Direction.UP: Direction.DOWN,
    Direction.DOWN: Direction.UP,
    Direction.LEFT: Direction.RIGHT,
    Direction.RIGHT: Direction.LEFT,
}


class SnakeGame:
    def __init__(self, width: int = 30, height: int = 20):
        self.width = width
        self.height = height
        self.score = 0
        self.alive = True
        self.direction = Direction.RIGHT

        mid_x, mid_y = width // 2, height // 2
        self.snake: deque[tuple[int, int]] = deque([
            (mid_x, mid_y),
            (mid_x - 1, mid_y),
            (mid_x - 2, mid_y),
        ])
        self.occupied: set[tuple[int, int]] = set(self.snake)
        self.food = self._place_food()

    def _place_food(self) -> tuple[int, int]:
        while True:
            pos = (random.randint(0, self.width - 1),
                   random.randint(0, self.height - 1))
            if pos not in self.occupied:
                return pos

    def change_direction(self, new_dir: Direction) -> None:
        if new_dir != OPPOSITE[self.direction]:
            self.direction = new_dir

    def tick(self) -> bool:
        if not self.alive:
            return False

        dx, dy = self.direction.value
        hx, hy = self.snake[0]
        head = (hx + dx, hy + dy)

        # Wall collision
        if not (0 <= head[0] < self.width and 0 <= head[1] < self.height):
            self.alive = False
            return False

        # Self collision
        if head in self.occupied:
            self.alive = False
            return False

        self.snake.appendleft(head)
        self.occupied.add(head)

        if head == self.food:
            self.score += 10
            self.food = self._place_food()
        else:
            tail = self.snake.pop()
            self.occupied.discard(tail)

        return True


def main(stdscr: curses.window) -> None:
    curses.curs_set(0)
    curses.start_color()
    curses.use_default_colors()
    curses.init_pair(1, curses.COLOR_GREEN, -1)   # snake
    curses.init_pair(2, curses.COLOR_RED, -1)      # food
    curses.init_pair(3, curses.COLOR_CYAN, -1)     # border
    curses.init_pair(4, curses.COLOR_YELLOW, -1)   # score

    game = SnakeGame()
    stdscr.timeout(110)

    key_map = {
        curses.KEY_UP: Direction.UP,       ord('w'): Direction.UP,
        curses.KEY_DOWN: Direction.DOWN,   ord('s'): Direction.DOWN,
        curses.KEY_LEFT: Direction.LEFT,   ord('a'): Direction.LEFT,
        curses.KEY_RIGHT: Direction.RIGHT, ord('d'): Direction.RIGHT,
    }

    while True:
        stdscr.erase()
        h, w = stdscr.getmaxyx()

        # Offset to center the board
        ox = max(0, (w - game.width * 2 - 2) // 2)
        oy = max(0, (h - game.height - 4) // 2)

        # Score
        score_text = f" SCORE: {game.score} "
        stdscr.addstr(oy, ox, score_text, curses.color_pair(4) | curses.A_BOLD)

        # Border
        border_w = game.width * 2 + 2
        stdscr.addstr(oy + 1, ox, "+" + "-" * (border_w - 2) + "+", curses.color_pair(3))
        for row in range(game.height):
            stdscr.addstr(oy + 2 + row, ox, "|", curses.color_pair(3))
            stdscr.addstr(oy + 2 + row, ox + border_w - 1, "|", curses.color_pair(3))
        stdscr.addstr(oy + 2 + game.height, ox, "+" + "-" * (border_w - 2) + "+", curses.color_pair(3))

        # Food
        fx, fy = game.food
        stdscr.addstr(oy + 2 + fy, ox + 1 + fx * 2, "@@", curses.color_pair(2) | curses.A_BOLD)

        # Snake
        for i, (sx, sy) in enumerate(game.snake):
            ch = "██" if i == 0 else "▓▓"
            stdscr.addstr(oy + 2 + sy, ox + 1 + sx * 2, ch, curses.color_pair(1))

        if not game.alive:
            msg = f"  GAME OVER — Score: {game.score}  Press 'r' to restart, 'q' to quit  "
            my = oy + 2 + game.height // 2
            mx = max(0, ox + (border_w - len(msg)) // 2)
            stdscr.addstr(my, mx, msg, curses.A_REVERSE | curses.A_BOLD)

        stdscr.refresh()

        key = stdscr.getch()

        if key == ord('q'):
            break
        if not game.alive:
            if key == ord('r'):
                game = SnakeGame()
            continue

        if key in key_map:
            game.change_direction(key_map[key])

        game.tick()


if __name__ == "__main__":
    curses.wrapper(main)
